docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,367 @@
+	Senders() []p2p.ID
+```
+- We introduce a new communication channel, called `Mempool Control Channel`. The channel
+is used to transmit the messages needed for the implementation of the protocol. The channel ID
+is `31` with priority `10` (for context, the mempool channel sending transaction has a priority of `5`).
+
+Additionally, the mempool reactor is extended with a 
+`gossipRouter` and a `redundancyControl` struct to keep track of the redundancy
+of transactions in the system. 
+Each node keeps track of the routes it disabled between its peers. 
+
+```go
+type gossipRouter struct {
+	mtx cmtsync.RWMutex
+	// A set of `source -> target` routes that are disabled for disseminating transactions. Source
+	// and target are node IDs.
+	disabledRoutes map[p2p.ID]p2pIDSet
+}
+``` 
+
+
+```go
+type redundancyControl struct {
+	txsPerAdjustment int64
+
+	// Pre-computed upper and lower bounds of accepted redundancy.
+	lowerBound float64
+	upperBound float64
+
+	mtx          cmtsync.RWMutex
+	firstTimeTxs int64 // number of transactions received for the first time
+	duplicates   int64 // number of duplicate transactions
+
+	// If true, do not send HaveTx messages.
+	blockHaveTx atomic.Bool
+}
+
+```
+
+
+#### Added logic to existing functions
+
+The overall behaviour of existing functions remains the same when DOG is disabled. 
+
+Enabling DOG causes the following changes in behaviour:
+
+`Receive`
+
+The function now handles messages from two channels. The `MempoolControlChannel`
+transmits `HaveTx` and `ResetRoute` messages that keep track of routes between peers. 
+
+`TryAddTx` 
+
+When DOG is enabled we count the number of unique and duplicate transactions, and send
+`HaveTx` messages if the redundancy is too high. 
+
+Once a `HaveTx` message is sent,
+sending further `HaveTx` messages is blocked. It is unblocked by the redundancy control mechanism 
+once it is triggered again. This will be discussed below, but overall the frequency
+of `HaveTx` messages should be aligned with the time it takes to send a message and 
+adjust routing between peers. 
+
+`broadcastTxRoutine`
+
+Before sending a transaction, we filter the peers to send to based on the `disabledRoutes`. Again,
+only if DOG is enabled. 
+
+`RemovePeer`
+
+When a peer is removed, we remove any existing entries related to this peer from the `disabledRoutes` 
+map and explicitly trigger redundancy re-adjustment. The later is not strictly needed, but 
+can lead to quicker propagation of this information through the network.
+
+Note that we had considered a scenario where, upon removing a peer, a node sends `ResetRoute` messages to all its peers. 
+This lead to routes being disabled too many times, and the process of stabilizing redundancy would be re-triggered
+too frequently. The adapted logic relies in the redundancy controller to decide whether a `ResetRoute` 
+message should be sent. 
+
+
+#### New p2p messages
```

Comment:
I would move this section that describes the new messages at the beginning of the Detailed Design section.